File: SortVision/public/code/bucket/python/bucketSort.py
"""Bucket Sort (for numbers in [0, 1)).

Time: Avg O(n + k), Worst O(n^2)
Space: O(n + k)
"""

def bucket_sort(arr: list[float]) -> list[float]:
    if len(arr) <= 1:
        return arr

    n = len(arr)
    buckets: list[list[float]] = [[] for _ in range(n)]

    for x in arr:
        idx = min(n - 1, int(x * n))
        buckets[idx].append(x)

    k = 0
    for bucket in buckets:
        bucket.sort()
        for x in bucket:
            arr[k] = x
            k += 1

    return arr


if __name__ == "__main__":
    arr = [0.42, 0.32, 0.23, 0.52, 0.25, 0.47, 0.51]
    print(bucket_sort(arr))

File: SortVision/public/code/bucket/swift/bucketSort.swift
import Foundation

// Bucket Sort (for numbers in [0, 1))
// Time: Avg O(n + k), Worst O(n^2), Space: O(n + k)
func bucketSort(_ arr: inout [Double]) {
    guard arr.count > 1 else { return }

    let n = arr.count
    var buckets = Array(repeating: [Double](), count: n)

    for x in arr {
        let idx = min(n - 1, Int(x * Double(n)))
        buckets[idx].append(x)
    }

    var k = 0
    for i in 0..<n {
        buckets[i].sort()
        for x in buckets[i] {
            arr[k] = x
            k += 1
        }
    }
}

var arr: [Double] = [0.42, 0.32, 0.23, 0.52, 0.25, 0.47, 0.51]
bucketSort(&arr)
print(arr)
